File: backend/src/main/java/com/hgw/gestionale/brand/controller/BrandController.java
package com.hgw.gestionale.brand.controller;

import com.hgw.gestionale.brand.dto.BrandResponse;
import com.hgw.gestionale.brand.dto.CreateBrandRequest;
import com.hgw.gestionale.brand.dto.UpdateBrandRequest;
import com.hgw.gestionale.brand.service.BrandService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.MediaType;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping(value = "/api/brands", produces = MediaType.APPLICATION_JSON_VALUE)
@RequiredArgsConstructor
@Slf4j
public class BrandController {
    private final BrandService brandService;

    @PostMapping
    public BrandResponse createBrand(@Valid @RequestBody CreateBrandRequest request) {
        log.info("BrandController.create creazione brand: {}", request.nome());
        return brandService.create(request);
    }

    @GetMapping
    public List<BrandResponse> getAllBrands() {
        log.info("BrandController.getAll richiesta lista brand");
        return brandService.getAll();
    }

    @GetMapping("/{id}")
    public BrandResponse getBrandById(@PathVariable Long id) {
        log.info("BrandController.getById richiesta brand id={}", id);
        return brandService.getById(id);
    }

    @GetMapping("/by-name/{nome}")
    public BrandResponse getBrandByName(@PathVariable String nome) {
        log.info("BrandController.getByName richiesta brand nome={}", nome);
        return brandService.getByName(nome);
    }

    @GetMapping("/search")
    public List<BrandResponse> searchBrand(@RequestParam String nome) {
        log.info("BrandController.search ricerca brand nome={}", nome);
        return brandService.search(nome);
    }

    @PutMapping("/{id}")
    public BrandResponse updateBrand(
            @PathVariable Long id,
            @Valid @RequestBody UpdateBrandRequest request
    ) {
        log.info("BrandController.update aggiornamento brand id={}", id);
        return brandService.update(id, request);
    }

    @DeleteMapping("/{id}")
    public void deleteBrand(@PathVariable Long id) {
        log.info("BrandController.delete eliminazione brand id={}", id);
        brandService.delete(id);
    }
}

File: backend-node/src/controller/brandController.ts
import {
    Body,
    Controller,
    Delete,
    Get,
    OperationId,
    Path,
    Post,
    Put,
    Query,
    Route,
    Security,
    Tags,
} from "tsoa";
import { BrandResponse, CreateBrandRequest, UpdateBrandRequest } from "../dto/brand.dto";
import { BrandService } from "../service/brand.service";

const brandService = new BrandService();

@Route("brands")
@Tags("Brands")
@Security("bearerAuth")
export class BrandController extends Controller {

    @Post("/")
    @OperationId("createBrand")
    public async createBrand(@Body() body: CreateBrandRequest): Promise<BrandResponse> {
        return brandService.create(body);
    }

    @Get("/")
    @OperationId("getAllBrands")
    public async getAllBrands(): Promise<BrandResponse[]> {
        return brandService.getAll();
    }

    @Get("/search")
    @OperationId("searchBrands")
    public async searchBrands(@Query() nome: string): Promise<BrandResponse[]> {
        return brandService.search(nome);
    }

    @Get("/by-name/{nome}")
    @OperationId("getBrandByName")
    public async getBrandByName(@Path() nome: string): Promise<BrandResponse> {
        return brandService.getByName(nome);
    }

    @Get("/{id}")
    @OperationId("getBrandById")
    public async getBrandById(@Path() id: number): Promise<BrandResponse> {
        return brandService.getById(id);
    }

    @Put("/{id}")
    @OperationId("updateBrand")
    public async updateBrand(
        @Path() id: number,
        @Body() body: UpdateBrandRequest
    ): Promise<BrandResponse> {
        return brandService.update(id, body);
    }

    @Delete("/{id}")
    @OperationId("deleteBrand")
    public async deleteBrand(@Path() id: number): Promise<void> {
        brandService.delete(id);
    }
}
